plugins/spectate/spectate.cpp

```diff
@@ -0,0 +1,367 @@
+        out.print("Spectate mode enabled!\n");
+        using namespace EM::EventType;
+        EM::EventHandler ticking(onTick, 15);
+        EM::EventHandler start(onJobStart, 0);
+        EM::EventHandler complete(onJobCompletion, 0);
+        EM::registerListener(EventType::TICK, ticking, plugin_self);
+        EM::registerListener(EventType::JOB_STARTED, start, plugin_self);
+        EM::registerListener(EventType::JOB_COMPLETED, complete, plugin_self);
+        enabled = true; // enable_auto_unpause won't do anything without this set now
+        enable_auto_unpause(out, unpause_enabled);
+    } else if (!enable && enabled) {
+        // warp 8, engage!
+        out.print("Spectate mode disabled!\n");
+        EM::unregisterAll(plugin_self);
+        // we need to retain whether auto-unpause is enabled, but we also need to disable its effect
+        bool temp = unpause_enabled;
+        enable_auto_unpause(out, false);
+        unpause_enabled = temp;
+        job_tracker.clear();
+        freq.clear();
+    }
+    enabled = enable;
+    return DFHack::CR_OK;
+}
+
+DFhackCExport command_result plugin_onstatechange(color_ostream &out, state_change_event event) {
+    if (enabled) {
+        switch (event) {
+            case SC_MAP_UNLOADED:
+            case SC_BEGIN_UNLOAD:
+            case SC_WORLD_UNLOADED:
+                our_dorf = nullptr;
+                job_watched = nullptr;
+                following_dwarf = false;
+            default:
+                break;
+        }
+    }
+    return CR_OK;
+}
+
+DFhackCExport command_result plugin_onupdate(color_ostream &out) {
+    // keeps announcement pause settings locked
+    World::Update(); // from pause.h
+    if (lock_collision) {
+        if (unpause_enabled) {
+            // player asked for auto-unpause enabled
+            World::SaveAnnouncementSettings();
+            if (World::DisableAnnouncementPausing()) {
+                // now that we've got what we want, we can lock it down
+                lock_collision = false;
+            }
+        } else {
+            if (World::RestoreAnnouncementSettings()) {
+                lock_collision = false;
+            }
+        }
+    }
+    int failsafe = 0;
+    while (unpause_enabled && !world->status.popups.empty() && ++failsafe <= 10) {
+        // dismiss announcement popup(s)
+        Gui::getCurViewscreen(true)->feed_key(interface_key::CLOSE_MEGA_ANNOUNCEMENT);
+        if (World::ReadPauseState()) {
+            // WARNING: This has a possibility of conflicting with `reveal hell` - if Hermes himself runs `reveal hell` on precisely the right moment that is
+            World::SetPauseState(false);
+        }
+    }
+    if (failsafe >= 10) {
+        out.printerr("spectate encountered a problem dismissing a popup!");
+    }
+    if (disengage_enabled && !World::ReadPauseState()) {
+        if (our_dorf && our_dorf->id != df::global::ui->follow_unit) {
+            plugin_enable(out, false);
+        }
```

Comment:
I wonder if auto pausing would be desirable to the folks who'd use this feature.
```suggestion
        if (our_dorf && our_dorf->id != df::global::ui->follow_unit) {
            plugin_enable(out, false);
            World::SetPauseState(true);
        }
```

Reply:
I don't think so -- follow doesn't do it, and I think we should mimic that behavior

Reply:
alright, sounds good